persist/sqldb/offload_node_status_repo.go

```diff
@@ -1,151 +1,174 @@
-		if errors.IsCode(CodeDBUpdateRowNotFound, err) {
-			return wdc.insert(wfdb)
-		} else {
-			log.Warn(err)
-			return errors.InternalErrorf("Error in inserting workflow in persistence. %v", err)
-		}
+		return "", "", err
 	}
 
-	log.Info("Workflow update successfully into persistence")
-	return nil
+	h := fnv.New32()
+	_, _ = h.Write(marshalled)
+	return string(marshalled), fmt.Sprintf("fnv:%v", h.Sum32()), nil
 }
 
-func (wdc *nodeOffloadRepo) insert(wfDB *WorkflowRecord) error {
-	tx, err := wdc.session.NewTx(context.TODO())
+func (wdc *nodeOffloadRepo) Save(uid, namespace string, nodes wfv1.Nodes) (string, error) {
+
+	marshalled, version, err := nodeStatusVersion(nodes)
 	if err != nil {
-		return errors.InternalErrorf("Error in creating transaction. %v", err)
+		return "", err
 	}
 
-	defer func() {
-		if tx != nil {
-			err := tx.Close()
-			if err != nil {
-				log.Warnf("Transaction failed to close")
-			}
-		}
-	}()
-
-	_, err = tx.Collection(wdc.tableName).Insert(wfDB)
-	if err != nil {
-		return errors.InternalErrorf("Error in inserting workflow in persistence. %v", err)
+	record := &nodesRecord{
+		ClusterName: wdc.clusterName,
+		UID:         uid,
+		Version:     version,
+		Namespace:   namespace,
+		Nodes:       marshalled,
 	}
 
-	err = tx.Commit()
+	logCtx := log.WithFields(log.Fields{"uid": uid, "version": version})
+	logCtx.Debug("Offloading nodes")
+	_, err = wdc.session.Collection(wdc.tableName).Insert(record)
 	if err != nil {
-		return errors.InternalErrorf("Error in Committing workflow insert in persistence. %v", err)
+		// if we have a duplicate, then it must have the same name+namespace+offloadVersion, which MUST mean that we
+		// have already written this record
+		if !strings.Contains(err.Error(), "duplicate key") {
+			return "", err
+		}
 	}
 
-	return nil
-}
-
-func (wdc *nodeOffloadRepo) update(wfDB *WorkflowRecord) error {
-	tx, err := wdc.session.NewTx(context.TODO())
+	logCtx.Info("Nodes offloaded, cleaning up old offloads")
+
+	// This might fail, which kind of fine (maybe a bug).
+	// It might not delete all records, which is also fine, as we always key on resource version.
+	// We also want to keep enough around so that we can service watches.
+	_, err = wdc.session.
+		DeleteFrom(wdc.tableName).
+		Where(db.Cond{"clustername": wdc.clusterName}).
+		And(db.Cond{"uid": uid}).
+		And(db.Cond{"version <>": version}).
+		And(db.Cond{"updatedat + interval '5' minute <": "now()"}).
+		Exec()
 	if err != nil {
-		return errors.InternalErrorf("Error in creating transaction. %v", err)
+		return "", err
 	}
+	return version, nil
+}
 
-	defer func() {
-		if tx != nil {
-			err := tx.Close()
-			if err != nil {
-				log.Warnf("Transaction failed to close")
-			}
-		}
-	}()
-
-	err = tx.Collection(wdc.tableName).UpdateReturning(wfDB)
+func (wdc *nodeOffloadRepo) Get(uid, version string) (wfv1.Nodes, error) {
+	log.WithFields(log.Fields{"uid": uid, "version": version}).Debug("Getting offloaded nodes")
```

Comment:
If the workflow has been updated in the Kubernetes API but the database hasn't been updated yet, then this could fail - should it retry for this case?

Reply:
I don't think so (but happy to be proved wrong). 

We use the version so can have historical versions of the record in the database. AFAIK, only the controller can change these, and we keep them for 5m (same as Etcd).

Reply:
Ah you're right. What if:

- Goroutine 1 reads workflow status nodes from DB
- Goroutine 1 updates status nodes object in memory

- Goroutine 2 reads workflow status nodes from DB (same workflow)
- Goroutine 2 updates status nodes object in memory

- Goroutine 1 writes updated status nodes to DB
- Goroutine 1 updates workflow (exc. status nodes)  in Kubernetes API
- Goroutine 2 writes updated status nodes to DB
- Goroutine 2 updates workflow (exc. status nodes) in Kubernetes API (note that this will call reapplyUpdate which will merge any Goroutine 1 changes from the Kubernetes API before persisting the workflow)

Won't the status nodes in the DB be missing the changes from Goroutine 1?

Reply:
@mark9white ,

Thank you for this - I really appreciate being challenged on this and it is critical we get it right.

> Goroutine 1 reads workflow status nodes from DB
> Goroutine 1 updates status nodes object in memory
> Goroutine 2 reads workflow status nodes from DB (same workflow)
> Goroutine 2 updates status nodes object in memory

At this point both routines have the same values for

- `metadata/resourceVersion`
-  `status/nodes`
- `status/offloadedNodeStatusVersion`

> Goroutine 1 writes updated status nodes to DB
> Goroutine 1 updates workflow (exc. status nodes) in Kubernetes API

At this point routine 2 has a different values for resourceVersion/nodes/offloadedNodeStatusVersion.

> Goroutine 2 writes updated status nodes to DB

Now we have two records in the database each with a different versions.

> Goroutine 2 updates workflow (exc. status nodes) in Kubernetes API (note that this will call reapplyUpdate which will merge any Goroutine 1 changes from the Kubernetes API before persisting the workflow)

This is correct . Aside -  I wonder if `reapplyUpdate` is safe anyway?

The K8S will have goroutine 2's offloadedNodeStatusVersion - there will be two records in the database with two version. The older one is effectively orphaned and only needed so that we can support watches.

At this point Goroutine has a problem. Regardless of what happens with node status, it has out of data information. This is the behaviour today.

Please keep this feedback coming!

Reply:
Agreed that this issue is already happening today already when offload nodes are turned on (though not when they're turned off as the status nodes will be part of the merge in reapplyUpdate).

I would think reapplyUpdate should be safe as long as there's no conflicts between the two updates. The safer way would be for goroutine 2 to abort and restart.